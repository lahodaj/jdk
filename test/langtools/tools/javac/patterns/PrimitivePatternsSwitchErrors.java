/*
 * @test /nodynamiccopyright/
 * @bug 8304487
 * @summary Compiler Implementation for Primitive types in patterns, instanceof, and switch (Preview)
 * @compile/fail/ref=PrimitivePatternsSwitchErrors.out --enable-preview -source ${jdk.version} -XDrawDiagnostics -XDshould-stop.at=FLOW PrimitivePatternsSwitchErrors.java
 */
public class PrimitivePatternsSwitchErrors {
    record R_int(int x) {}

    public static void dominationBetweenPrimitivePatterns() {
        int i = 42;
        switch (i) {
            case short s -> System.out.println("its a short");
            case byte b  -> System.out.println("its a byte"); // Error - dominated!
            default      -> System.out.println("any other integral value");
        }
    }

    public static int dominationWithRecordPatterns() {
        R_int r = new R_int(42);
        return switch (r) {
            case R_int(int x) -> 1;
            case R_int(byte x) -> 2;  // Error - dominated!
        };
    }

    public static int inconvertibleNestedComponent() {
        R_int r = new R_int(42);
        return switch (r) {
            case R_int(Long x) -> 1; // inconvertible
        };
    }

    public static int nonExhaustive1() {
        int i = 42;
        return switch (i) {  // Error - not exhaustive
            case short s -> s;
        };
    }

    public static int nonExhaustive2() {
        int i = 42;
        return switch (i) { // Error - not exhaustive
            case byte  b -> 1;
            case short s -> 2;
        };
    }

    public static int nonExhaustive3() {
        int i = 42;
        return switch (i) { // Error - not exhaustive
            case byte  b -> 1;
            case float f -> 2;
        };
    }

    public static int dominationBetweenBoxedAndPrimitive() {
        int i = 42;
        return switch (i) {
            case Integer ib  -> ib;
            case byte ip     -> ip; // Error - dominated!
        };
    }

    public static int constantDominatedWithPrimitivePattern() {
        int i = 42;
        return switch (i) {
            case int j -> 42;
            case 43    -> -1;   // Error - dominated!
        };
    }

    public static int constantDominatedWithFloatPrimitivePattern() {
        float f = 42.0f;
        return switch (f) {
            case Float ff -> 42;
            case 43.0f    -> -1;   // Error - dominated!
        };
    }

    void switchLongOverByte(byte b) {
        switch (b) {
            case 0L: return ;
        }
    }

    void switchOverPrimitiveFloatFromInt(float f) {
        switch (f) {
            case 16777216:
                break;
            case 16777217:
                break;
            default:
                break;
        }
    }

    void switchOverNotRepresentableFloat(Float f) {
        switch (f) {
            case 1.0f:
                break;
            case 0.999999999f:
                break;
            case Float fi:
                break;
        }
    }

    int switchOverPrimitiveBooleanExhaustiveWithNonPermittedDefault(boolean b) {
        return switch (b) {
            case true -> 1;
            case false -> 2;
            default -> 3;
        };
    }
}
