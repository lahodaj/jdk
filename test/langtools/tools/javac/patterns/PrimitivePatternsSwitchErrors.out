PrimitivePatternsSwitchErrors.java:14:18: compiler.err.pattern.dominated
PrimitivePatternsSwitchErrors.java:23:18: compiler.err.pattern.dominated
PrimitivePatternsSwitchErrors.java:30:24: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: int, java.lang.Long)
PrimitivePatternsSwitchErrors.java:61:18: compiler.err.pattern.dominated
PrimitivePatternsSwitchErrors.java:69:18: compiler.err.pattern.dominated
PrimitivePatternsSwitchErrors.java:77:18: compiler.err.pattern.dominated
PrimitivePatternsSwitchErrors.java:83:18: compiler.err.prob.found.req: (compiler.misc.possible.loss.of.precision: long, byte)
PrimitivePatternsSwitchErrors.java:89:18: compiler.err.constant.label.not.compatible: int, float
PrimitivePatternsSwitchErrors.java:91:18: compiler.err.constant.label.not.compatible: int, float
PrimitivePatternsSwitchErrors.java:102:13: compiler.err.duplicate.case.label
PrimitivePatternsSwitchErrors.java:29:16: compiler.err.not.exhaustive
PrimitivePatternsSwitchErrors.java:36:16: compiler.err.not.exhaustive
PrimitivePatternsSwitchErrors.java:43:16: compiler.err.not.exhaustive
PrimitivePatternsSwitchErrors.java:51:16: compiler.err.not.exhaustive
PrimitivePatternsSwitchErrors.java:110:16: compiler.err.default.label.not.allowed
- compiler.note.preview.filename: PrimitivePatternsSwitchErrors.java, DEFAULT
- compiler.note.preview.recompile
15 errors
